### PR TITLE
[Feature] Add watch command

### PR DESCRIPTION
**Pull request by rygine**
The `watch` command will watch for changes in your package folders and run a script when it detects a change. It expect at least 2 arguments: `[glob]` and `[command]`.

Example: `lerna watch 'src/**/*.js' build`

In the example above, `npm run build <path/to/changed/file>` will run in the package folder of the file that changed. By default, the relative path to the changed file will be passed as the first argument of the command. You may specify additional arguments to include when running the watch command.

Example: `lerna watch 'src/**/*.js' test arg1 arg2 arg3`

This will run `npm run test <path/to/changed/file> arg1 arg2 arg3`.

### CODE REVIEW

**Comment by rygine** [comment]
A few things to note:

* the colorized output when a file changes is up for debate
* unit tests aren't included as i don't think they are needed
* when an error occurs, the watcher is stopped and the error is passed to the callback
* the first argument is the relative path of the changed file, this could be an absolute path

**Comment by joscha** [comment]
can't this be done via `lerna exec -- pwd | sed 1,2d` and passing the returned folders to chokidar?

**Comment by thejameskyle** [comment]
ehh– you don't have a watch task running to see whenever you need to run `npm install` I don't see a need for this

**Comment by rygine** [comment]
@thejameskyle the watch command here is intended to run the build command within packages as i'm making changes. this is currently not possible with babel. sure, i can watch files in 1 package and have them build, but not all packages at the same time. am i missing something?

this way, i'm free to work on multiple dependent packages at the same time and have them rebuild and hot reload into my webpack dev server. i find this highly useful.

i never intended the watcher to run `npm install` and don't see a use case for that.

**Comment by doug-wade** [comment]
Since `npm install` calls `prepublish` internally, you could use `npm install` to run the build in watch mode (and check dependencies as a side-effect).

**Comment by rygine** [comment]
@doug-wade i don't understand your use case. if all of my modules are bootstrapped, and i want to start watching for changes at an arbitrary time, how do i accomplish that?

**Comment by doug-wade** [comment]
@rygine Consider [this commit](https://github.com/redfin/react-server/pull/302/commits/be2be2ab93dfda2c699c80645c02c85b23167d9b) that didn't merge to React Server.  Initially, as part of `prepublish`, I babelified the sources of my babel plugin, and then exported the transpiled javascript as `main`.  @thejameskyle says that we don't need a watch task to see when we need to run `npm install`, but in this case, a `watch` mode that would detect changes to `src` and run `npm install` because [`npm install` runs `prepublish`](https://docs.npmjs.com/misc/scripts#description).  With your implementation, it would make more sense to use `watch` to run `npm run prepublish`, but even running `npm install` would be a huge improvement.

I've been implementing watch tasks as scripts, and then using `lerna exec -- npm run watch`, which works pretty okay, but is a lot of boilerplate.

**Comment by rygine** [comment]
@doug-wade running `npm install` after a source file change makes absolutely no sense to me. i have no idea why it was even mentioned as a use case here. this is simply for transpiling changed source files so that it's picked up by a dev server during development.

if i'm working on `packageA` which depends on `packageB` and `packageC`, i'd want to see the realtime changes of `packageB' or `packageC`. with this watcher, my webpack setup would pick up the changes and hot reload them for me.

note: the `main` entry for all my packages points to a transpiled version in the `/dist` folder.

**Comment by paulyoung** [comment]
A `lerna watch` command seems incredibly useful.

I tried `lerna exec -- npm run watch` and it seemed to get stuck on the first package since the `watch` command doesn't exit.

In my case I would probably set `"prepublish": "npm run clean; npm run build"` and then run `lerna watch 'src/**/*.js' prepublish`.

I'm currently manually building individual packages when I make changes so that other dependent packages can get the latest. This is sometimes several levels deep.

If there's a way to achieve the same thing without these changes, that would be great and I'd love to know how.

Perhaps I'm missing something, but I'm afraid I don't at all follow why `npm install` was mentioned.

**Comment by thejameskyle** [comment]
> the watch command here is intended to run the build command within packages as i'm making changes

I missed this, when I first read this I thought it was a watch version of `lerna bootstrap`. I blame having to wake up at 5:30am this morning

**Comment by joscha** [comment]
I like the idea of watching, but it seems to me as if this does not necessarily need to be a first-class citizen of lerna, as it can be composed, e.g.
`lerna exec -- pwd | sed 1,2d | sed -e 's/$/\/**\/*.js/' | xargs npm run watch -- --files`

**Comment by rygine** [comment]
@joscha is lerna not all about convenience and ease of use? i believe the `clean` command is a product of this thinking. :) while i'm sure some bash magic can achieve a similar end, it can look rather confusing to those not familiar with the command line. not to mention, the above command may not work on all platforms, creating a need to conjure up another way to achieve the same task.

how are you currently handling file changes during development?

i've only rarely used `sed` and in the most basic way. a quick googling tells me it's not something you master in a few minutes. i was unable to get your command to do anything. i'm sure it's something on my end.

**Comment by joscha** [comment]
I get the convenience part ;) I think for me mass file watching is a relict of the past in most cases. Modern build tools would allow you to pass some entry points and watch the whole dependency graph, no matter what exact location those files are in - globs and directory thinking make refactors harder, so I prefer an `import { x } from 'module'` (that module gets then added to the watched graph) over an explicit `npm run watch module/**/*.js`.

**Comment by rygine** [comment]
@thejameskyle @hzoo @gigabo @joscha is this still something to be considered for lerna right now? i know there's a push to gets things merged so that #174 can be implemented. in which case,  i can close this now and it can come back later as `lerna-command-watch`.

**Comment by nruhe** [comment]
I would definitely like to know the status on this. We have major need of the ability to run our prepublish script after changes so that the requiring repos can get the latest content in dist.

**Comment by rygine** [comment]
@nruhe that's exactly what my team is doing with this command. i just rebased. i'd love to get more feedback from others.

**Comment by gigabo** [comment]
Yeah, this actually seems really sweet to me.

Our workflow on React Server involves a lot of:

```bash
$ cd ../<other-package> && npm run prepublish && cd - && npm start
```

This would really help streamline that.

> i still need to add documentation.

And tests? 😁

**Comment by rygine** [comment]
yes, tests too!

i'm thinking the only thing i could test is the value of `this.packagesToWatch` for packages that don't have the script. everything else seems to be covered.

File: src/commands/WatchCommand.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      const [pkg] = path.relative(this.repository.packagesLocation, src).split(path.sep);
+      const packageLocation = path.join(this.repository.packagesLocation, pkg);
+      const srcFile = path.relative(packageLocation, src);
+      console.log(`${chalk.cyan(`${pkg}/${srcFile}`)} changed, running ${chalk.yellow(this.script)} script`);
+      NpmUtilities.runScriptInDir(this.script, [srcFile].concat(this.args), packageLocation, (err, stdout) => {
+        if (err) {
+          this.logger.error(err);
+        } else {
+          this.logger.info(stdout);
+        }
```
Comment (gigabo) [inline src/commands/WatchCommand.js]: Probably good to get `stdout` on error here, too.

Can just follow the pattern [you established](https://github.com/lerna/lerna/pull/265/files) in `RunCommand.js`.

File: src/commands/WatchCommand.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  execute() {
+    this.logger.info(`Watching for changes in package files that match '${this.glob}'`);
+    const globPath = (pkg) => path.join(this.repository.packagesLocation, pkg.name, this.glob);
+    // watch all package files matching passed glob
+    chokidar.watch(this.packagesToWatch.map(globPath)).on("change", (src) => {
+      const [pkg] = path.relative(this.repository.packagesLocation, src).split(path.sep);
+      const packageLocation = path.join(this.repository.packagesLocation, pkg);
+      const srcFile = path.relative(packageLocation, src);
+      console.log(`${chalk.cyan(`${pkg}/${srcFile}`)} changed, running ${chalk.yellow(this.script)} script`);
+      NpmUtilities.runScriptInDir(this.script, [srcFile].concat(this.args), packageLocation, (err, stdout) => {
```
Comment (gigabo) [inline src/commands/WatchCommand.js]: Hmm... does it always make sense to automatically pass `srcFile` as an argument?

What if I'm just doing `lerna watch src/* prepublish`?

Comment (rygine) [inline src/commands/WatchCommand.js]: Good point. My thought was that `srcFile` would just be ignored if it's not needed. In my environment, I transpile single files as they are changed as Webpack freaks out when you remove and replace the `/dist` folder while it's running.

Would it make sense to have a flag for this?

Here's the watch command I run:

```
lerna watch 'src/**/*.@(js|css)' build -- '-- --incremental'
```

I know it looks weird. We have a tool that wraps `babel` and by default, it cleans `/dist`, then transpiles everything in `/src` to `/dist`. The `incremental` flag take a source file argument and just transpiles that.

File: src/commands/WatchCommand.js (hunk 1 above)
Comment (rygine) [inline src/commands/WatchCommand.js]: Good catch!

File: src/commands/WatchCommand.js (hunk 2 above)
Comment (gigabo) [inline src/commands/WatchCommand.js]: Yeah, the file argument breaks if it's not expected:

image

One option would be to make the file available as an environment variable, similar to [this](https://github.com/lerna/lerna/pull/276/files).

So if I wanted to use it I could say:

```bash
lerna watch src/* eslint -- --fix $LERNA_CHANGED_FILE
```

That would also allow custom placement of the source file argument _within_ the arguments array, so it's not just forced to be at the beginning.

**Comment by rygine** [comment]
Should the `watch` command also take flag values from `lerna.json`?

```js
{
  "lerna": "2.0.0-beta.26",
  "version": "1.0.0",
  "watchConfig": {
    "ignore": "my-ignored-component"
  }
}
```

**Comment by gigabo** [comment]
Maybe?  Maybe add it if someone requests?

**Comment by rygine** [comment]
@thejameskyle does this feature fit into the scope of what Lerna is trying to do?

**Comment by thejameskyle** [comment]
Yeah, the addition of watch tools makes it a bit heavy of a dependency but that could be easily fixed with in the long term plan I had for Lerna having a global CLI that calls out to an installed package (https://github.com/lerna/lerna/issues/138). The global cli can have all the watch tools installed globally and be used in tandem. Will just require some refactoring later.

**Comment by nruhe** [comment]
What's the ETA on merging this?

**Comment by nruhe** [comment]
Sorry to keep pushing on this one, but we're in the process of adopting a monorepo and we -really- need the ability to have the system run build/prepublish commands on changes (e.g. npm prepublish). @thejameskyle can you comment on whether this will be added any time soon or if we should find an alternative?

**Comment by gigabo** [comment]
@rygine Want to dust this off? 😁

**Comment by rygine** [comment]
sure thing. will take a look in the next few days.

**Comment by nruhe** [comment]
Thank you very much. :)

**Comment by rygine** [comment]
@nruhe @gigabo i've updated the PR. i _think_ it's good to go. all tests seem to pass.

@gigabo i don't have a test for when the `ignore` and `scope` flags are set. i didn't see this being tested in other commands that support both. is this something we should be testing?

**Comment by gigabo** [comment]
> i don't have a test for when the `ignore` and `scope` flags are set.

Always nice to have more tests, but I don't think we need to hold this PR up for that.

**Comment by reggi** [comment]
Just tried to run this and got an error :(

```
npm i https://github.com/awakenetworks/lerna.git#a251ead423c77fc8e13185f70a052eaa8d6adf93 -g
```

```
Thomass-MBP-2:wonder thomas$ lerna
module.js:472
    throw err;
    ^

Error: Cannot find module '../lib/index'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/thomas/.nvm/versions/node/v7.2.1/lib/node_modules/lerna/bin/lerna.js:3:13)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
```

**Comment by rygine** [comment]
@gigabo the AppVeyor failure seems to be completely unrelated. :/

**Comment by reggi** [comment]
This might be useful! https://github.com/mateodelnorte/loop-things

**Comment by doug-wade** [comment]
@rygine the appveyor failure is likely unrelated.  We've [dropped the version of node that failed](https://github.com/lerna/lerna/commit/79b29a30a5b60f3701d59c8ccd70e21febe38d43) -- wanna rebase and try again?

File: bin/lerna.js [hunk 3]
```diff
@@ -10,25 +10,26 @@ var cli = meow([
...
+  "  publish                Publish updated packages to npm",
+  "  updated                Check which packages have changed since the last release",
+  "  import                 Import a package with git history from an external repository",
+  "  clean                  Remove the node_modules directory from all packages",
+  "  diff                   Diff all packages or a single package since the last release",
+  "  init                   Initialize a lerna repo",
+  "  run                    Run npm script in each package",
+  "  exec                   Run a command in each package",
+  "  ls                     List all public packages",
+  "  watch [glob] [script]  Watch for changes to package files and run a script when a change is detected",
```
Comment (evocateur) [inline bin/lerna.js]: Oy, we really have to get away from this fiddly manual formatting. (urge to refactor into [yargs](https://github.com/yargs/yargs/) rising...)

None of the other commands list their positional arguments here, can we stay consistent with that until we move to a better argparser?

**Comment by Bnaya** [comment]
Maybe i've missed something in all of the discussion,
But:
Why not add support running npm tasks in the packages in parallel? 
say, each package has it own `watch` task, and in the task you have `babel -w`

lerna exec claims to run tasks concurrently, but as i see it block on the first package watch command

**Comment by gigabo** [comment]
@Bnaya Yep, that's an alternative way to get at good watch behavior in a Lerna repo.

There's some discussion about that here: https://github.com/lerna/lerna/issues/446.

**Pull request closed without merging**